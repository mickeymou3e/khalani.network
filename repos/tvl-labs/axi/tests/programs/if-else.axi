function p(x: Nat) -> Boolean;
function q(x: Nat) -> Boolean;
function g(x: Nat) -> Nat;

function f(x: Nat) -> Nat {
    if p(x) {
        match x {
            Succ(y) => if q(y) { g(y) } else { f(y) },
            Zero => Zero
        }
    } else { Zero }
}

!claim(forall y: Nat. (
    p(Succ(y)) ==>
    q(y) ==>
    f(Succ(y)) = g(y)
));

!claim(forall y: Nat. (
    p(Succ(y)) ==>
    ~q(y) ==>
    f(Succ(y)) = f(y)
));

!claim(forall x: Nat. (
    ~p(x) ==>
    f(x) = Zero
));
