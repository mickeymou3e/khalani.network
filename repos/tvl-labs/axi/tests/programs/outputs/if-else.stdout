Result: Nothing
Result: Nothing
Result: Nothing
Result: Nothing
Result: Proved(forall Nat. if(p[Succ[@0:Nat]):Nat], if(q[@0:Nat], =[f[Succ[@0:Nat]):Nat]):Nat, g[@0:Nat]):Nat])))
Result: Proved(forall Nat. if(p[Succ[@0:Nat]):Nat], if(not(q[@0:Nat]), =[f[Succ[@0:Nat]):Nat]):Nat, f[@0:Nat]):Nat])))
Result: Proved(forall Nat. if(not(p[@0:Nat]), =[f[@0:Nat]):Nat, Zero:Nat]))
